### PR TITLE
adding CRUD apis

### PR DESCRIPTION
**Pull request by kunal9724kg**

### CODE REVIEW

File: functions/routes/categoryRoutes.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+function getAllCategories(req, res) {
+    categoriesRef
+        .get()
+        .then( (categories) => {
+            return res.status(200).json({
+                success: true,
+                message: 'List of All Categories Received Successfully',
+                data: {
+                    categories: categories.docs.map(doc => doc.data())
```
Comment (ciphereck) [inline functions/routes/categoryRoutes.js:24]: need to check if categories is null or undefined?

File: functions/routes/categoryRoutes.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            return res.status(500).json({
+                success: false,
+                message: error
+            })
+        });
+}
+
+function addNewCategory(req,res) {
+    categoriesRef
+        .doc(req.body.id)
```
Comment (ciphereck) [inline functions/routes/categoryRoutes.js:38]: check if req.body.id is null or undefined.

Comment (ciphereck) [reply functions/routes/categoryRoutes.js:38]: use req.body.category.id instead of req.body.id

File: functions/routes/categoryRoutes.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        .set(req.body)
```
Comment (ciphereck) [inline functions/routes/categoryRoutes.js:39]: use req.body.category

File: functions/routes/categoryRoutes.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        .then( () => {
+            
+            return res.status(200).json({
+                success: true,
+                message: 'New Category Added Successfully'
```
Comment (ciphereck) [inline functions/routes/categoryRoutes.js:44]: New Category?
Maybe give name of category or id of category.

File: functions/routes/categoryRoutes.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            return res.status(500).json({
+                success: false,
+                message: error
+            })
+        });
+}
+
+function updateCategory(req, res) {
+    categoriesRef
+    .doc(req.body.id)
```
Comment (ciphereck) [inline functions/routes/categoryRoutes.js:57]: tab before .doc and everything else?

File: functions/routes/categoryRoutes.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    .update(req.body)
+    .then( () => {
+        return res.status(200).json({
+            success: true,
+            message: 'Category Updated Successfully'
```
Comment (ciphereck) [inline functions/routes/categoryRoutes.js:62]: Category id updated successfully.

File: functions/routes/categoryRoutes.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        return res.status(500).json({
+            success: false,
+            message: error
+        })        
+    });
+}
+
+function deleteCategory(req, res) {
+    categoriesRef
+    .doc(req.body.id)
```
Comment (ciphereck) [inline functions/routes/categoryRoutes.js:75]: tabs after before .doc and everything else?

Comment (ciphereck) [inline functions/routes/categoryRoutes.js:75]: 1. use req.body.category.id
2. check whether id is null (and undefined) or not?

File: functions/routes/categoryRoutes.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    .delete()
+    .then( () => {
+        return res.status(200).json({
+            success: true,
+            message: 'Category Deleted Successfully'
```
Comment (ciphereck) [inline functions/routes/categoryRoutes.js:80]: Category id deleted successfully if exist.

File: functions/index.js [hunk 9]
```diff
@@ -1,9 +1,25 @@
 const admin = require('firebase-admin');
 const functions = require('firebase-functions');
 const express = require('express');
-
 const app = express();
+const bodyParser = require('body-parser');
```
Comment (ankurcharan) [inline functions/index.js:5]: all require statements in the beginning

File: functions/routes/categoryRoutes.js (hunk 2 above)
Comment (ankurcharan) [reply functions/routes/categoryRoutes.js:38]: category.id is too much nesting for a single parameter.
Null/Undefined checks should be there

File: functions/routes/categoryRoutes.js (hunk 4 above)
Comment (ankurcharan) [reply functions/routes/categoryRoutes.js:44]: Generate the message string dynamic 
Ex. 'Clothes Category Added'

File: functions/routes/categoryRoutes.js (hunk 5 above)
Comment (ankurcharan) [reply functions/routes/categoryRoutes.js:57]: indentation is not required here

File: functions/routes/categoryRoutes.js (hunk 8 above)
Comment (ankurcharan) [reply functions/routes/categoryRoutes.js:80]: I guess firebase will handle the if exist part.
@kunal9724kg?

Ex. ''Clothes Deleted Successfully

File: functions/routes/productRoutes.js [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+function addNewproduct(req,res) {
+    productsRef
+        .doc(req.body.id)
+        .set(req.body)
+        .then( () => {    
+            return res.status(200).json({
+                success: true,
+                message: 'New Product Added Successfully'
```
Comment (ankurcharan) [inline functions/routes/productRoutes.js:43]: name the product

**Comment by kunal9724kg** [comment]
Review the latest changes

File: functions/index.js [hunk 11]
```diff
@@ -1,9 +1,26 @@
...
 const app = express();
 
+app.use(bodyParser.urlencoded({extended:false}));
+
+var serviceAccount = require("../ecommerce-eed77-firebase-adminsdk-dikkn-01149939c5.json");
+
+admin.initializeApp({
+  credential: admin.credential.cert(serviceAccount),
+  databaseURL: "https://ecommerce-eed77.firebaseio.com"
```
Comment (ciphereck) [inline functions/index.js:13]: is it possible to store database url in some config file and same for serviceAccount?

Not Important Right Now

File: functions/model/Product.js [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+        if(product === undefined){
+            throw new Error('Send Product Details');
+        }
+        if(product.id === undefined){
+            throw new Error('Send Product Id');
+        }
+        
+        this.id = product.id;
+        this.name = product.name;
+        this.price = product.price;
```
Comment (ciphereck) [inline functions/model/Product.js:12]: Do we need some checks here?
Like price must not be null or name must not be null?

Again Not Important

File: functions/model/SubCategory.js [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        if(subCategory === undefined){
+            throw new Error('Send Sub-Category');
+        }
+        if(subCategory.id === undefined){
+            throw new Error('Send Sub-Category Id');
+        }
+
+        this.id = subCategory.id
+        this.name = subCategory.name
+        this.categoryId = subCategory.categoryId
```
Comment (ciphereck) [inline functions/model/SubCategory.js:12]: Category Id must not be Null

Important

File: functions/routes/categoryRoutes.js [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+        .then( (categories) => {
+            return res.status(200).json({
+                success: true,
+                message: 'List of All Categories Received Successfully',
+                data: {
+                    categories: categories
+                                    .docs
+                                    .map(doc => Category
+                                        .getCategoryConverter()
+                                        .fromFirestore(doc))
```
Comment (ciphereck) [inline functions/routes/categoryRoutes.js:29]: Later we can put a try catch block here too...
To check whether categoryDoc get converted into Category Class or not.

Not important Right Now

File: functions/routes/categoryRoutes.js [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+        return res.status(400).json({
+			success: false,
+			message: error.message
+		});
+    }
+
+    categoriesRef
+        .doc(category.id)
+        .withConverter(Category.getCategoryConverter())
+        .set(Object.assign({},category))
```
Comment (ciphereck) [inline functions/routes/categoryRoutes.js:55]: extra space here after .assign{},

Object.assign({}, category)

File: functions/routes/categoryRoutes.js [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+        })
+        .catch( (error) => {
+            return res.status(500).json({
+                success: false,
+                message: error.message
+            }) 
+        });
+}
+
+
```
Comment (ciphereck) [inline functions/routes/categoryRoutes.js:128]: why extra lines?

File: functions/routes/productRoutes.js [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+        })
+        .catch( (error) => {
+            return res.status(500).json({
+                success: false,
+                message: error
+            }) 
+        });
+}
+
+
```
Comment (ciphereck) [inline functions/routes/productRoutes.js:126]: same comments as on category
And remove extra space later

File: functions/routes/subCategoryRoutes.js [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+const SubCategory = require('../model/SubCategory')
+
+                            
+router
+    .route('/')
+    .get(getAllSubCategories)
+    .post(addNewSubCategory)
+    .put(updateSubCategory)
+    .delete(deleteSubCategory);
+
```
Comment (ciphereck) [inline functions/routes/subCategoryRoutes.js:15]: extra line?

File: functions/routes/subCategoryRoutes.js [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+        })
+        .catch( (error) => {
+            return res.status(500).json({
+                success: false,
+                message: error
+            }) 
+        });
+}
+
+
```
Comment (ciphereck) [inline functions/routes/subCategoryRoutes.js:129]: extra space?

**Pull request merged**